src/closures/closure_operators.jl

```diff
@@ -0,0 +1,532 @@
+∂y_aca(i, j, k, grid, ϕ::AbstractArray) = δy_c2f(grid, ϕ, i, j, k) / grid.Δy
+∂z_aac(i, j, k, grid, ϕ::AbstractArray) = δz_c2f(grid, ϕ, i, j, k) / grid.Δz
+
+#
+# Differential operators for functions
+#
+
+"""
+    ∂x_caa(i, j, k, grid, F, args...)
+
+Differentiate the function or callable object
+
+    `F(i, j, k, grid, args...)`
+
+located at `caa` in `x`, across `faa`.
+"""
+function ∂x_caa(i, j, k, grid, F::Function, args...)
+    i⁻ = decmod1(i, grid.Nx)
+    return (F(i, j, k, grid, args...) - F(i⁻, j, k, grid, args...)) / grid.Δx
+end
+
+"""
+    ∂x_faa(i, j, k, grid, F, args...)
+
+Differentiate the function or callable object
+
+    `F(i, j, k, grid, args...)`
+
+located at `faa` in `x`, across `caa`.
+"""
+function ∂x_faa(i, j, k, grid, F::Function, args...)
+    i⁺ = incmod1(i, grid.Nx)
+    return (F(i⁺, j, k, grid, args...) - F(i, j, k, grid, args...)) / grid.Δx
+end
+
+"""
+    ∂y_aca(i, j, k, grid, F, args...)
+
+Differentiate the function or callable object
+
+    `F(i, j, k, grid, args...)`
+
+located at `aca` in `y`, across `afa`.
+"""
+function ∂y_aca(i, j, k, grid, F::Function, args...)
+    j⁻ = decmod1(j, grid.Ny)
+    return (F(i, j, k, grid, args...) - F(i, j⁻, k, grid, args...)) / grid.Δy
+end
+
+"""
+    ∂y_afa(i, j, k, grid, F, args...)
+
+Differentiate the function or callable object
+
+    `F(i, j, k, grid, args...)`
+
+located at `afa` in `y`, across `aca`.
+"""
+function ∂y_afa(i, j, k, grid, F::Function, args...)
+    j⁺ = incmod1(j, grid.Ny)
+    return (F(i, j⁺, k, grid, args...) - F(i, j, k, grid, args...)) / grid.Δy
+end
+
+"""
+    ∂z_aac(i, j, k, grid, F, args...)
+
+Differentiate the function or callable object
+
+    `F(i, j, k, grid, args...)`
+
+located at `aac` in `z`, across `aaf`.
+"""
+function ∂z_aac(i, j, k, grid::Grid{T}, F::Function, args...) where T
+    if k < 2
+        return -zero(T)
+    else
+        return (F(i, j, k-1, grid, args...) - F(i, j, k, grid, args...)) / grid.Δz
+    end
+end
+
+"""
+    ∂z_aaf(i, j, k, grid, F, args...)
+
+Differentiate the function or callable object
+
+    `F(i, j, k, grid, args...)`
+
+located at `aaf` in `z`, across `aac`.
+"""
+function ∂z_aaf(i, j, k, grid::Grid{T}, F::Function, args...) where T
+    if k >= grid.Nz
+        return -zero(T)
+    else
+        return (F(i, j, k, grid, args...) - F(i, j, k+1, grid, args...)) / grid.Δz
+    end
+end
+
+#
+# Double differentiation
+#
+
+∂x²_faa(i, j, k, grid, u) = ∂x_caa(i, j, k, grid, ∂x_faa, u)
+∂x²_caa(i, j, k, grid, ϕ) = ∂x_faa(i, j, k, grid, ∂x_caa, ϕ)
+
+∂y²_afa(i, j, k, grid, v) = ∂y_aca(i, j, k, grid, ∂y_afa, v)
+∂y²_aca(i, j, k, grid, ϕ) = ∂y_afa(i, j, k, grid, ∂y_aca, ϕ)
+
+∂z²_aaf(i, j, k, grid, w) = ∂z_aac(i, j, k, grid, ∂z_aaf, w)
+∂z²_aac(i, j, k, grid, ϕ) = ∂z_aaf(i, j, k, grid, ∂z_aac, ϕ)
+
+#
+# Interpolation operations for functions
+#
+
+"""
+    ▶x_caa(i, j, k, grid, F, args...)
+
+Interpolate the function or callable object
+
+    `F(i, j, k, grid, args...)`
+
+from `caa` to `faa`."
```

Comment:
After thinking about I think you have a good point. Let's talk about it before making a final decision as this will take a long time to change with a high potential for bugs.